### ISSUE TITLE
Build MS2DeepScore model class

### ISSUE THREAD

**Issue by florian-huber** [matchms/ms2deepscore#15]
Build keras model class, see [here](https://keras.io/api/models/model/)

**Comment by florian-huber** [comment]
@svenvanderburg, I forgot to put the model here too. 

One I used was this:
```python
dense_dims = [600, 500, 500, 400]
dropout_rate = 0.25

# Encoder network
input_data  = Input(shape = (vector_dim), name="encoder_input")
encoding = Dense(dense_dims[0], activation = 'relu', #kernel_regularizer=tf.keras.regularizers.l1(1e-6),
                 name="dense_101")(input_data)
encoding = BatchNormalization()(encoding)
encoding = Dropout(dropout_rate)(encoding)
encoding = Dense(dense_dims[1], activation = 'relu', name="dense_102")(encoding)
encoding = BatchNormalization()(encoding)
encoding = Dropout(dropout_rate)(encoding)
encoding = Dense(dense_dims[2], activation = 'relu', name="dense_103")(encoding)
encoding = BatchNormalization()(encoding)
encoding = Dropout(dropout_rate)(encoding)
encoding = Dense(dense_dims[3], activation = 'relu', name="dense_104")(encoding)

encoder = Model(input_data, encoding)

# Scoring network
spectrum1 = Input(shape = (vector_dim), name="scoring_input1")
spectrum2 = Input(shape = (vector_dim), name="scoring_input2")

embedding1 = encoder(spectrum1)
embedding2 = encoder(spectrum2)

cosine_similarity = Dot(axes=(1, 1), normalize=True, name="cosine_similarity")([embedding1, embedding2])

scoring_network = Model(inputs=[spectrum1, spectrum2], outputs = [cosine_similarity])

scoring_network.summary()
encoder.summary()
```

**Comment by florian-huber** [comment]
And training was done like this:
```python
epochs = 150
learning_rate = 0.002 #0.001
metrics = ["mae"]

from tensorflow.keras.optimizers import Adam
from tensorflow.keras.callbacks import EarlyStopping, ModelCheckpoint, TensorBoard
from datetime import datetime
timestamp = datetime.now().strftime("%Y%m%d-%H%M%S")

# Parameters
patience_scoring_net = 10
scoring_network_output_file = "scoring_network_201123_All_cosine_medium_fixedBins.hdf5"

scoring_network.compile(
    loss='mse',
    optimizer=Adam(lr=learning_rate),
    metrics=metrics)

checkpointer = ModelCheckpoint(
    filepath = scoring_network_output_file,
    monitor='val_loss', mode="min",
    verbose=1,
    save_best_only=True
    )

earlystopper_scoring_net = EarlyStopping(
    monitor='val_loss', mode="min",
    patience=patience_scoring_net,
    verbose=1
    )

history = scoring_network.fit_generator(
    generator=train_generator,
    validation_data=val_generator,
    epochs = epochs,
    callbacks = [
        earlystopper_scoring_net,
        checkpointer,
        ]
    )
```

**Comment by florian-huber** [comment]
Related to failing tests in #16 I realized that `model.fit_generator` is deprecated, so I tried running 
```python
model.fit(train_generator, validation_data=val_generator, ...)
```
which seems to work --> PR #18

**Comment by florian-huber** [comment]
I also started using the new code to build a typical workflow. I then realized that `SpectrumBinner` should somehow be tied to the model itself to make sure that a user will import a **model** together with the **vocabulary** and **transformation settings**(bin sizes etc) that were used to train the model.

We could add the `SpectrumBinner` to the model class so that a user could load a pretrained model and use it like this:
```python
model = load_siamese_model(filename)
model.summary()
print(f"no. of known peaks: {len(model.spectrum_binner.known_bins)})")

# and after importing the spectrums to use the model on:
input_spectrums_binned = model.spectrum_binner.transform(input_spectrums)
```

**Comment by svenvanderburg** [comment]
> Related to failing tests in #16 I realized that `model.fit_generator` is deprecated, so I tried running
> 
> ```python
> model.fit(train_generator, validation_data=val_generator, ...)
> ```
> 
> which seems to work --> PR #18

Cool!

> I also started using the new code to build a typical workflow. I then realized that `SpectrumBinner` should somehow be tied to the model itself to make sure that a user will import a **model** together with the **vocabulary** and **transformation settings**(bin sizes etc) that were used to train the model.
> 
> We could add the `SpectrumBinner` to the model class so that a user could load a pretrained model and use it like this:
> 
> ```python
> model = load_siamese_model(filename)
> model.summary()
> print(f"no. of known peaks: {len(model.spectrum_binner.known_bins)})")
> 
> # and after importing the spectrums to use the model on:
> input_spectrums_binned = model.spectrum_binner.transform(input_spectrums)
> ```
> 
> Let me know what you think of this @svenvanderburg .

Yes, that makes a lot of sense. That would be part of the high-level #17 wrapper right? Not of the model class introduced in this issue.

**Comment by svenvanderburg** [comment]
This issue was fixed in #16, but I'll keep it open until the above discussion is resolved.

**Comment by florian-huber** [comment]
Let's continue the above discussion in #23.
The model class itself was added with PR #16, so I'll close this issue.
